---
title: "Data generation and processing"
subtitle: "Sequencing protocols, best practice variant calling and filtering"
author:
    - "Per Unneberg"
date: "`r format(Sys.time(), '%d %B, %Y')`"
institute: NBIS
from: markdown+emoji
format:
  revealjs:
    theme:
      - white
      - ../nbis.scss
    self-contained: true
    toc: true
    toc-depth: 1
    slide-level: 3
    slide-number: true
    preview-links: true
    chalkboard: false
    footer: Data processing
    logo: https://nbis.se/assets/img/logos/nbislogo-green.svg
    smaller: true
    highlight-style: gruvbox
    fig-align: center
    width: 1050
    height: 700
    margin: 0.1
    navigation-mode: vertical
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/apa.csl
bibliography: ../resources/bibliography.bib
execute:
  echo: true
  warning: false
  cache: true
  include: true
  autodep: true
  eval: true
  error: true
opts:
  knitr:
    code-fold: true
    tidy: true
    fig-format: svg
---

## Setup  {visibility="hidden" .unnumbered .unlisted}

```{r libs}
#| echo: false
#| eval: true
#| cache: FALSE
library(readxl)
library(curl)
library(ggplot2)
library(scales)
library(viridis)
bw <- theme_bw(base_size=24) %+replace% theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
theme_set(bw)
knitr::knit_hooks$set(inline = function(x) {
                      prettyNum(x, big.mark=" ")
                  })
```

# Data generation

### Population genomics - the data

Since the goal of population genomics is to analyze variation in a set
of individuals, data generation consists of compiling variation data
from individuals. Here the focus is on next-generation sequencing
data.


```{r }
#| label: fig-sequencing-cost
#| echo: false
#| eval: true
#| fig-cap: Sequencing cost ($) per megabase [@_DNASequencingCosts_]
url <- "https://www.genome.gov/sites/default/files/media/files/2021-11/Sequencing_Cost_Data_Table_Aug2021.xls"
if (!file.exists(basename(url)))
    download.file(url, destfile=basename(url))
data <- read_excel(basename(url))
colnames(data) <- c("Date", "Mb", "Genome")
ggplot(data, aes(x=as.Date(Date), y=Mb)) + geom_line() + xlab("") + ylab("") + scale_y_continuous(trans="log10", labels=comma, n.breaks=6) + scale_x_date(breaks="year", labels=date_format("%Y"))
```



###


###

![@lou_BeginnerGuideLowcoverage_2021](../img/sequencing_approaches_lou_2021.png)


### RADSeq


### High coverage sequencing data

Pros: call genotypes with confidence

cons: cost

::: {.notes }

[@lou_BeginnerGuideLowcoverage_2021 p.5967] points out that "full
genome coverage is often needed to understand mechanisms of
adaptation" but that high-coverage sequencing still too expensive

:::


### Low coverage sequencing data

lcWGS - alleviates cost issue



# Read mapping and variant calling



### GATK best practice

Point out: not optimal for

### Alternative variant callers

#### freebayes


#### bcftools


#### ANGSd


### PoolSeq

#### Popoolation

### Refs {.unnumbered .unlisted}
::: { #refs }
:::
