---
title: "Population genetics"
subtitle: "Foundations"
author:
    - "Per Unneberg"
date: "`r format(Sys.time(), '%d %B, %Y')`"
institute: NBIS
from: markdown+emoji
format:
  revealjs:
    theme:
      - white
      - ../nbis.scss
    self-contained: true
    toc: true
    toc-depth: 1
    slide-level: 3
    slide-number: true
    preview-links: true
    chalkboard: false
    footer: Population genetics foundations
    logo: https://nbis.se/assets/img/logos/nbislogo-green.svg
    smaller: true
    highlight-style: gruvbox
    fig-height: 12
    fig-width: 10
    navigation-mode: vertical
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/apa.csl
execute:
  echo: true
  warning: false
  cache: true
  include: true
  autodep: true
  eval: true
  error: true
opts:
  knitr:
    code-fold: true
    tidy: true
    fig-format: svg
---

### Setup  {visibility="hidden" .unnumbered .unlisted}

```{r libs}
#| echo: false
#| eval: true
#| cache: false
library(knitr)
library(ggplot2)
library(viridis)
library(dotenv)
library(igraph)
library(gganimate)
library(patchwork)
library(tidyr)
bw <- theme_bw(base_size=18) %+replace% theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
theme_set(bw)
knitr::knit_hooks$set(inline = function(x) {
                      prettyNum(x, big.mark=" ")
                  })
```
```{r }
#| label: utility-functions
#| echo: false
#| eval: true
#| cache: false
```


::: {.notes }



:::


# Population genetics

### DNA variation



::: {.dna }

| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
| T | T | A | C | A | A | T | C | C | G  | A  | T  | C  | G  | T  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| T | C | A | C | A | A | T | G | C | G  | A  | T  | G  | G  | A  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

The main data for molecular population genetics are DNA sequences. The
alignment above shows a *sample* of four DNA *sequences*. Each
sequence has 15 nucleotides (*sites*) "from the same *locus*
(location) on a chromosome" [p.2
@hahn_MolecularPopulationGenetics_2019]

::: {.fragment}

Alternative names for *sequence*:

:::

::: {.incremental}

- *chromosome*
- *gene*
- *allele* (different by origin)
- *sample*
- *cistron*

:::

::: {.fragment}

We will preferentially use *sequence* or *chromosome* to refer to an
entire sequence, and *allele* to refer to individual nucleotides that
differ.

:::

### DNA variation - monomorphic sites

::: {.dna }

| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
| T | T | A | C | A | A | T | C | C | G  | A  | T  | C  | G  | T  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| T | C | A | C | A | A | T | G | C | G  | A  | T  | G  | G  | A  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| * |   | * | * |   | * | * |   | * | *  |    | *  |    | *  | T  |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

The alignment has 4 DNA sequences where each sequence has length
$L=15$. A site where all nucleotides (alleles) are identical is called
a *monomorphic* site (indicated with asterisks above). There are 9
monomorphic sites.



### DNA variation - segregating sites

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

A site where there are different nucleotides (alleles) is called a
*segregating* site (indicated with asterisks above), often denoted
*S*. There are $S=6$ segregating sites.


::: {.fragment}

Alternative names for *segregating site* are:

:::

::: {.incremental}

- *polymorphism*
- *mutation*
- *single nucleotide polymorphism (SNP)*

:::


::: {.fragment}

*mutation* may refer to the process that generates new variation and
the new variants generated by this process, which is the way we will
use the term.


In contrast to *mutation* which corresponds to within-species
variation, a *substitution* refers to DNA differences between species.

:::

### DNA variation - major and minor alleles

::: {.dna }

| 1 | 2     | 3 | 4 | 5     | 6 | 7 | 8     | 9 | 10 | 11    | 12 | 13    | 14 | 15    |
|:-:|:-----:|:-:|:-:|:-----:|:-:|:-:|:-----:|:-:|:--:|:-----:|:--:|:-----:|:--:|:-----:|
| T | T     | A | C | A     | A | T | **C** | C | G  | A     | T  | C     | G  | T     |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
| T | **C** | A | C | A     | A | T | G     | C | G  | A     | T  | **G** | G  | **A** |
| T | T     | A | C | **G** | A | T | G     | C | G  | **C** | T  | C     | G  | T     |
|   | *     |   |   | *     |   |   | *     |   |    | *     |    | *     |    | *     |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]"}

:::

Much of the nucleotide variation we study consists of *bi-allelic*
SNPs. The most common variant is called the *major* allele, and the
least common the *minor* allele.


The set of alleles found on a single sequence is called *haplotype*.



### Describing DNA variation


::: {.dna }

| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
| T | T | A | C | A | A | T | C | C | G  | A  | T  | C  | G  | T  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| T | C | A | C | A | A | T | G | C | G  | A  | T  | G  | G  | A  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,25]"}

:::

Once we have a sample of sequences we want to describe the observed
variation. At any position the *i*th allele has sample frequency
$p_i$, where the sum of all allele frequencies is 1. For instance, at
site 1, $p_T=1$ (and by extension $p_A=p_C=p_G=0$), and at site 2
$p_C=1/4$ and $p_T=3/4$.

#### Heterozygosity

The *heterozygosity* at a site is given by

$$
h = \frac{n}{n-1}\left(1 - \sum p_i^2\right)
$$

::: {.fragment}

#### Exercise: calculate the heterozygosity at sites 1 and 2

:::

::: {.fragment}

$$
h_1 = \frac{4}{3} \left(1 - p_T^2 \right) = 0 \\
h_2 = \frac{4}{3} \left(1 - \left(p_C^2 + p_T^2\right) \right) = \frac{4}{3} \left( 1 - \left(\frac{1}{16} + \frac{9}{16}\right)\right) = \frac{1}{2}
$$
:::


### Describing DNA variation


::: {.dna }

| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
| T | T | A | C | A | A | T | C | C | G  | A  | T  | C  | G  | T  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |
| T | C | A | C | A | A | T | G | C | G  | A  | T  | G  | G  | A  |
| T | T | A | C | G | A | T | G | C | G  | C  | T  | C  | G  | T  |

: {tbl-colwidths="[5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,25]"}

:::

#### Nucleotide diversity $\pi$

The *nucleotide diversity* is the sum of site heterozygosities:

$$
\pi = \sum_{j=1}^S h_j
$$

::: {.fragment}

#### Calculate the nucleotide diversity

:::

::: {.fragment}

Observation: $h_i$ either 1/2 or 2/3 (for sites with
$p_{major}=p_{minor}$).

:::

::: {.fragment}

$$
\pi = \frac{1}{2} + \frac{2}{3} + \frac{1}{2} + \frac{2}{3} + \frac{1}{2} + \frac{1}{2} = 3\frac{1}{3}
$$

:::

::: {.fragment}

Often we provide $\pi$ per site:

$$
\pi = 3.33/15 = 0.222
$$

:::

# Origin and change of variation


### Wright-Fisher model

:::: {.columns}

::: {.column width="50%"}


Model of evolution that describes sampling of alleles in a population
under the following assumptions:

::: {.incremental}

- no selection
- no mutation
- no migration
- non-overlapping generations
- random mating

:::


:::

::: {.column width="50%"}

:::

::::
### Wright-Fisher model

:::: {.columns}

::: {.column width="50%"}

Model of evolution that describes sampling of alleles in a population
under the following assumptions:

- no selection
- no mutation
- no migration
- non-overlapping generations
- random mating

:::

::: {.column width="50%"}


```{r }
#| label: wright-fisher-model-graph
#| echo: false
#| eval: true
wf_model_igraph <- function(width=10, height=10, parents_first=FALSE, untangled=TRUE) {
    G <- list()
    g <- make_empty_graph()
    g <- add_vertices(g, width * height, color=NA, size=0, label=NA, frame.color=NA)
    g$perm <- unlist(lapply(seq(vcount(g), 1, - width), function(x) {seq(x - width + 1, x)}))
    g$layout <- layout_on_grid(g, width=width, height=height)
    i <- 1
    for (i in seq(1, height)) {
        n0 <- (i - 1) * width + 1
        n1 <- i * width
        if (i >= 2) {
            p0 <- (i - 2) * width + 1
            p1 <- (i - 1) * width
            parents <- sample(p0:p1, width, replace=TRUE)
            if (untangled)
                parents <- sort(parents)
            children <- seq((i - 1) * width + 1, i * width)
            edges <- unlist(Map(c, parents, children))
            V(g)[parents]$color <- "black"
            G[[length(G) + 1]] <- g
            g <- add_edges(g, edges, color="black", width=2, arrow.size=0)
        }
        V(g)[n0:n1]$frame.color <- "black"
        V(g)[n0:n1]$color <- NA
        V(g)[n0:n1]$size <- 3
        G[[length(G) + 1]] <- g
    }
    G
}
G <- wf_model_igraph()
```


```{r }
#| label: fig-wf-1
#| echo: false
#| eval: true
#| fig-height: 7
#| fig-width: 7
plot(permute(G[[1]], G[[1]]$perm))
text(x=0, y=1.2, paste("Starting population"), cex=2)
```


:::

::::

### Wright-Fisher model

:::: {.columns}

::: {.column width="50%"}

Model of evolution that describes sampling of alleles in a population
under the following assumptions:


- no selection
- no mutation
- no migration
- non-overlapping generations
- random mating



:::

::: {.column width="50%"}

```{r }
#| label: fig-wf-2
#| echo: false
#| eval: true
#| fig-height: 7
#| fig-width: 7
plot(permute(G[[2]], G[[1]]$perm))
text(x=0, y=1.2, paste("Sampled parents"), cex=2)
```


:::

::::

### Wright-Fisher model

:::: {.columns}

::: {.column width="50%"}

Model of evolution that describes sampling of alleles in a population
under the following assumptions:


- no selection
- no mutation
- no migration
- non-overlapping generations
- random mating



:::

::: {.column width="50%"}

```{r }
#| label: fig-wf-3
#| echo: false
#| eval: true
#| fig-height: 7
#| fig-width: 7
plot(permute(G[[3]], G[[1]]$perm))
text(x=0, y=1.2, paste("Offspring"), cex=2)
```


:::

::::

### Wright-Fisher model

:::: {.columns}

::: {.column width="30%"}


Introduce model as way to understand drift. Show n=10 and
interactively/iteratively build next generation(s). Forward
simulation. (Add pgip code)


:::

::: {.column width="70%"}


```{r }
#| label: wright-fisher-graphical-model
#| echo: false
#| eval: true
#| fig-show: animate
#| fig-keep: all
#| fig-align: center
#| fig-height: 7
#| fig-width: 7
for (i in seq(1, length(G))) {
    plot(permute(G[[i]], G[[1]]$perm))
}
```


:::

::::



::: {.notes }

See https://mikeyharper.uk/animated-plots-with-r/

:::

### Genetic drift

:::: {.columns}

::: {.column width="50%"}



:::

::: {.column width="50%"}

```{r }
#| label: wright-fisher-example
#| echo: true
#| eval: true
wright_fisher <- function(p0, N, iter) {
    x <- vector(mode="numeric", length=iter)
    x[1] <- p0
    for (i in seq(2, length(x)))
        x[i] <- rbinom(1, size=N, prob=x[i-1]) / N
    x
}
```


:::

::::



```{r }
#| label: genetic-drift-example
#| echo: false
#| eval: true
N <- c(10, 100, 200, 500, 1000, 1e4)
iter <- 500
p0 <- 0.2
N <- rep(10, 5)

wf <- as.data.frame(do.call("cbind", lapply(N, function(n) {wright_fisher(p0, n, iter)})))
colnames(wf) <- N
wf$generation <- seq(1, iter)
data <- pivot_longer(wf, cols=as.character(N))
colnames(data) <- c("generation", "N", "frequency")
data$N <- factor(data$N, levels=as.character(N))
```

```{r }
#| label: fig-genetic-drift-N-100
#| echo: false
#| eval: true
ggplot(subset(data, data$N=="100"), aes(x=generation, y=frequency, color=N)) + geom_line(size=1) + xlim(1, 40) + ylim(0, 1)
```



Classic graph of allele: start at say 0.2 and drift with different Ne.
Point out that variation becomes smaller with large Ne, but that
eventually, an allele is fixed or lost (formula).

Mention *neutral* here as this is how alleles behave under drift: like
gas particle diffusing up and down


::: {.notes }

See alternative way of visualizing variation faceting on initial
frequency (showing P(fixation)) and population size (showing less
variance))

https://stephens999.github.io/fiveMinuteStats/wright_fisher_model.html

:::



### Mutation

Drift randomly "moves" frequencies, but once lost, how does new
variation arise? Mutation.

Introduce variation at rate $2N_e\mu$

Mutation - drift balance

Infinite sites vs infinite alleles

### Selection

Relative fitness.

Define s, positive and negative selection

### Direct selection

d_N/d_S


### Two-locus dynamics

Graph focal locus and flanking loci (Charlesworth figure)

- Linkage disequilibrium and recombination (each probably need
  separate treatment)
- Contrast linked selection with direct selection
- Genetic draft / background selection


### Neutral theory

Move right after drift?

- Hahn: $\mu = \nu*f_0$
- why population size matters
- Fig 4.2, Eq 4.2 from Lynch


### Nearly neutral theory

- Distribution of fitness effects?


# A primer on simulation

If simulation section show brief code examples

### Wright-Fisher model again

Repeat WF model. Point out that it can be used for *forward*
simulation. Show how many individuals are lost each generation (limit
e-1): lose lots of compute resources on the way to extant sample.

Solution: look backwards

### The coalescent

Introduce concept. Describe algorithm (show code - include as
exercise?). Genealogy and mutations:

- sim_ancestry
- sim_mutations

Note that #mutations propto branch length -> don't actually even need
to know the mutations (e.g. fastsimcoal)

### The coalescent: diminishing returns

Show plot of diminishing returns (better to add sites, not samples)

### The coalescent and diversity

Recalculate example

### Evolutionary processes and genealogies

Non-neutral processes change topologies in ways that we detect when
applying tests [@ferretti_DecomposingSiteFrequency_2017]

genealogies change due to non
Show some qualitative examples of how trees change for

- bottleneck

### msprime


### SLiM

Mention advances in forward simulations


### Recapitation

Combine the best of two worlds

### Recipes

Show some recipes and the pgip CLI
