---
title: "Population genomics in practice"
subtitle: "What is population genomics?"
author:
    - "Per Unneberg"
format:
  revealjs:
    slide-level: 2
    toc: false
---

## Setup  {visibility="hidden" .unnumbered .unlisted}

{{< include ../_knitr.qmd >}}

{{< include ../_rlibs.qmd >}}

```{r libs}
#| echo: false
#| eval: true
#| cache: false
library(readxl)
library(curl)
```

::: {.notes }

Points from [@hahn_MolecularPopulationGenetics_2019]:

- whole-genome data instead of single loci - population genomics is
  population genetics for whole-genome sequences
  - if only this, not too exciting
- major promise: enables analyses not possible for single loci or that
  require genomic context
- addresses interactions between different forces, notably selection
  and demographic history

Some applications:

- genome-wide scans for selection
  - selection vs demography
- methods for genome-wide scans
- non-independence
  - different statistics rely on similar input
  - overlapping peaks from different statistics *not* independent

[@hartl1997principles p. 469-470]

[@li_InferenceHumanPopulation_2011, supplementary notes, p. 6]:

"...highly consistent except for the very recent history,
demonstrating the power of using whole-genome data."

:::

## Example: Population genetics of the coral *Acropora millepora*

Motivation: corals are facing hard times and to prevent future losses
of coral cover a better understanding of genetics is warranted.

<!-- markdownlint-disable MD013 -->

:::: {.columns}

::: {.column width="50%"}

::: {.fragment}

![Genome assembly and sampling](https://www.science.org/cms/10.1126/science.aba4674/asset/31658ca3-933a-4047-81e1-e900fd9a0bae/assets/graphic/369_aba4674_f1.jpeg){#fig-corals-assembly height=420}

:::

:::

::: {.column width="50%"}

::: {.fragment}

![Variation and demographic history](https://www.science.org/cms/10.1126/science.aba4674/asset/661d50bb-a085-41af-abdb-06c487ed4e63/assets/graphic/369_aba4674_f2.jpeg){#fig-corals-variation height=420}

:::

:::

::::

<!-- markdownlint-enable MD013 -->

::: { .flushright .smallest}

@fuller_PopulationGeneticsCoral_2020

:::

::: {.notes }

[@fuller_PopulationGeneticsCoral_2020] is an example of a population
genomics study that applies methods that could be seen as a basic
foundation of population genomics. We believe these present a minimum
toolkit of methods that should be known to anyone starting out in
population genomics, and that is sufficiently small to be presented in
a one-week workshop. At the end of this lecture, we will discuss some
more advanced applications in population genomics.

Walk through steps of the main figures which represent common analyses
in a population genomics experiment. No need to supply text; better to
exemplify via the figures.

1. genome assembly and sampling. Points to consider:
   - choice of reference individual
   - the number of populations
   - the number of samples (more sites better than many samples per
     population)
   - the geographical distribution of samples
   - sequencing depth (low-coverage often sufficient)

2. variation and demographic history
   - why do we want to do this?

:::

## Example: Population genetics of the coral *Acropora millepora*

Motivation: corals are facing hard times and to prevent future losses
of coral cover a better understanding of genetics is warranted.

<!-- markdownlint-disable MD013 -->

:::: {.columns}

::: {.column width="50%"}

![Population structure](https://www.science.org/cms/10.1126/science.aba4674/asset/4625c5a6-1e71-4b9f-9589-0fed7ab72d3d/assets/graphic/369_aba4674_f3.jpeg){#fig-corals-structure}

:::

::: {.column width="50%"}

::: {.fragment}

![Selection scan](https://www.science.org/cms/10.1126/science.aba4674/asset/83167ee1-c1f3-4ef8-88b3-d69a0cb93062/assets/graphic/369_aba4674_f4.jpeg){#fig-corals-selection height=420}

:::

:::

::::

<!-- markdownlint-enable MD013 -->

::: { .flushright .smallest}

@fuller_PopulationGeneticsCoral_2020

:::

## From population genetics to population genomics

<!-- markdownlint-disable MD013 -->

![The variable sites at the *Drosophila melanogaster* ADH locus
[@kreitman_NucleotidePolymorphismAlcohol_1983]](assets/images/kreitman-1983-table1.png)

<!-- markdownlint-enable MD013 -->

::: {.notes }

Population genetics focuses on the genetic basis of evolution. It is
mainly a theoretical subject, owing to the slow changes of genetic
variation. As such, it tries to explain the shape and structure of
genetic variation from theoretical predictions and models.

- from locus-based studies (e.g., alcohol dehydrogenase in Drosophila
  [@kreitman_NucleotidePolymorphismAlcohol_1983]) to genome-wide
  (e.g., Drosophila population genomics
  [@begun_PopulationGenomicsWholeGenome_2007]
- note: studied loci have often not been randomly chosen, which is
  another argument for whole-genome studies
- enabler: sequencing technology

[@fuller_PopulationGeneticsCoral_2020] paper has population genetics
in title -> population genetics is a key ingredient.

Refer to Hahn's points about learning something about global patterns:

- selection acts locally, demography globally
- the structure of genetic variation and how it depends on
  - recombination landscapes and linked selection
  - demographic changes
  - identification of neutral loci

So, not simply about applying 10000 selection tests for multiple loci

Population genetics in title of previous slide => population genomics
is large-scale population genetics? At the very least, shows the
foundational role population genetics plays.

All of the points above point to the importance of statistics which
implies mathematics / computational skills important

:::

## From population genetics to population genomics

![Patterns of polymorphism and divergence [@begun_PopulationGenomicsWholeGenome_2007]](https://journals.plos.org/plosbiology/article/figure/image?size=large&download=&id=10.1371/journal.pbio.0050310.g001)

## From population genetics to population genomics

![Numbers of polymorphic and fixed variants [@begun_PopulationGenomicsWholeGenome_2007]](https://journals.plos.org/plosbiology/article/figure/image?size=large&download=&id=10.1371/journal.pbio.0050310.t002)

## The technological revolution in **sequencing** and **computing**

:::: {.columns}

::: {.column width="50%"}

```{r }
#| label: fig-sequencing-cost
#| echo: false
#| eval: true
#| fig-cap: Sequencing cost ($) per megabase [@wetterstrandka_DNASequencingCosts_]
#| out-width: 400
url <- "https://www.genome.gov/sites/default/files/media/files/2021-11/Sequencing_Cost_Data_Table_Aug2021.xls"
if (!file.exists(basename(url))) {
  download.file(url, destfile = basename(url))
}
data <- read_excel(basename(url))
colnames(data) <- c("Date", "Mb", "Genome")
ggplot(data, aes(x = as.Date(Date), y = Mb)) +
  geom_line() +
  xlab("") +
  ylab("") +
  scale_y_continuous(trans = "log10", labels = comma, n.breaks = 6) +
  scale_x_date(breaks = "year", labels = date_format("%Y"))
```

:::

::: {.column width="50%"}

![Moore's law](https://ourworldindata.org/uploads/2022/03/Transistor-Count-over-time.png)

:::

::::

## Statistical inference in population genomics

The data deluge requires advanced statistical methods and models to do
inference. Today data production outpaces theoretical advances.
Therefore, take care not to attach too much faith to a test that
explains data well.

::: {.fragment}

A population genomics study should aim at generating a baseline model
that takes into account the processes that shape genetic variation
[@johri_RecommendationsImprovingStatistical_2022]:

1. mutation
2. recombination
3. gene conversion
4. purifying selection acting on functional regions and its effects on
   linked variants (background selection)
5. genetic drift with demographic history and geographic structure

:::

# Population genomics in the future

## Machine learning and AI for population genomics

## Ancestral recombination graph methods

- deep population sequencing
- argweaver
- tskit
- Relate
- ARG-Needle

## Ancient DNA

- Relate

## Coalescent and Wright-Fisher simulations

Use to determine significance in genome-wide scans

## Structural variation in natural populations

## Bibliography {.unnumbered .unlisted}

::: { #refs }
:::

:::::::::: {.notes }

- A brief history of population genetics
  - Mendel's peas
  - Galton, Pearson and Bateson: Of blending and particulate
    inheritance (the latter championed by Bateson)
  - Morgan's flies
  - Fisher and natural selection
  - The modern synthesis
  - Wright and drift
  - Kimura and the diffusion approximation - Follow
    [@casillas_MolecularPopulationGenetics_2017] history?
  - The allozyme era (FIXME: picture of allozyme gel)
    - Development of genetic diversity measures
      - *H* (*gene diversity*, *heterozygosity*): average proportion
        heterozygous loci
      - *P* (*gene polymorphism*): average proportion polymorphic loci
  - The nucleotide diversity era
  - The population genomics era

::::::::::
