---
format: html
engine: knitr
---

{{< include ../_knitr.qmd >}}

{{< include ../_rlibs.qmd >}}


```{r}
#| label: r-custom
#| echo: false
#| eval: true
#| warning: false
#| error: false
#| cache: false
library(readxl)
library(curl)
```



```{bash }
#| label: E210-data-setup
#| echo: false
#| eval: true
#if [ ! -e assets/data ]; then
#		mkdir -p assets/data
#fi
DATADIR=${QUARTO_PROJECT_DIR}/data/data/ooa-outgroups
#for f in ${DATADIR}/[CY]*.chr21.bam*; do
#    cp $f assets/data/$(basename $f)
#done

#cp $DATADIR/ooa.chr22.ref.fa assets/data/
```


```{r }
#| label: r-set-working-dir
#| echo: false
#| eval: true
knitr::opts_knit$set(root.dir = normalizePath(file.path("assets/data")))
knitr::opts_knit$get("root.dir")
```

```{r }
#| label: r-pwd
#| echo: true
#| eval: true
getwd()
```

# Genetic diversity {#sec-ex-genetic-diversity}

## Nucleotide diversity {#sec-ex-nucleotide-diversity}

::: {.callout-warning}

FIXME: pen-and-paper exercise to calculate diversity based on "fixed"
genotype calls and genotype likelihoods to highlight differences; the
latter requires GLs for **all** sites. Use Table 4.1 in
[@hartl1997principles, p.173] and/or Figure 3.1 in
[@hahn_MolecularPopulationGenetics_2019, p. 45]

:::

::: {.callout-note}

- use sequence masks (0/1) encoded gz files to enable quick calculations of feature-based stats (e.g. 4d-degenerate sites)
- use pixy to compare to vcf with all sites present


:::


FIXME: comparison of pixy and coverage-based filtering approach
